File: frontend/src/utils/logTransactions.js
import { toPaise, fromPaise } from './money.js';

export const DEFAULT_ACCOUNTS = ['Cash', 'UPI', 'Bank', 'Credit Card', 'Wallet'].map((name) => ({
  id: `account_${name.toLowerCase().replaceAll(' ', '_')}`,
  name,
  archived: false,
}));
export const DEFAULT_CATEGORIES = [
  'Food',
  'Transport',
  'Shopping',
  'Entertainment',
  'Bills',
  'Health',
  'Education',
  'Travel',
  'Fuel',
  'Subscriptions',
  'Rent',
  'Groceries',
  'Other',
].map((name) => ({ id: `category_${name.toLowerCase()}`, name, archived: false }));
export const transactionKind = (entry) => entry.transactionType || 'expense';
export const entryPaise = (entry) =>
  Number.isSafeInteger(entry.amountPaise) ? entry.amountPaise : toPaise(Number(entry.amount || 0));
export const localDateTime = (value = new Date()) => {
  const date = new Date(value);
  if (Number.isNaN(date.getTime())) return localDateTime();
  return new Date(date.getTime() - date.getTimezoneOffset() * 60000).toISOString().slice(0, 16);
};
export const transactionFields = (data) => {
  const amountPaise = toPaise(data.amount);
  const kind = data.transactionType || 'expense';
  if (!['expense', 'income', 'transfer'].includes(kind))
    throw new Error('Choose a transaction type.');
  if (amountPaise <= 0 || amountPaise > 100000000)
    throw new Error('Enter an amount between ₹0.01 and ₹10,00,000.');
  const accountId = data.accountId || 'account_cash';
  if (kind === 'transfer' && (!data.toAccountId || accountId === data.toAccountId))
    throw new Error('Choose two different accounts for a transfer.');
  if ((data.currency || 'INR') !== 'INR')
    throw new Error('Only INR entry is supported at present.');
  return {
    amountPaise,
    amount: fromPaise(amountPaise),
    currency: data.currency || 'INR',
    transactionType: kind,
    accountId,
    accountName: data.accountName || 'Cash',
    categoryId: data.categoryId || `category_${(data.category || 'Other').toLowerCase()}`,
    toAccountId: kind === 'transfer' ? data.toAccountId : '',
    toAccountName: kind === 'transfer' ? data.toAccountName || '' : '',
    friendId: data.friendId || '',
    friendName: data.friendName || '',
  };
};

export const evaluateMathExpression = (expr) => {
  if (typeof expr !== 'string') return expr;
  const clean = expr.trim();
  if (!clean) return '';
  if (/^-?\d+(\.\d+)?$/.test(clean)) return clean;
  if (!/^[0-9\s.+\-*/]+$/.test(clean)) return expr;
  try {
    const tokens = clean.match(/(\d+(?:\.\d+)?|[+\-*/])/g);
    if (!tokens || tokens.length <= 1) return clean;
    const values = [];
    const ops = [];
    for (let i = 0; i < tokens.length; i++) {
      const token = tokens[i];
      if (['+', '-', '*', '/'].includes(token)) {
        ops.push(token);
      } else {
        let num = parseFloat(token);
        if (Number.isNaN(num)) return expr;
        if (ops.length > 0 && (ops[ops.length - 1] === '*' || ops[ops.length - 1] === '/')) {
          const op = ops.pop();
          const prev = values.pop();
          num = op === '*' ? prev * num : num !== 0 ? prev / num : prev;
        }
        values.push(num);
      }
    }
    if (values.length !== ops.length + 1) return expr;
    let result = values[0];
    for (let i = 0; i < ops.length; i++) {
      const op = ops[i];
      const next = values[i + 1];
      if (next === undefined) break;
      if (op === '+') result += next;
      else if (op === '-') result -= next;
    }
    if (Number.isFinite(result) && result >= 0) {
      return (Math.round(result * 100) / 100).toFixed(2);
    }
    return expr;
  } catch {
    return expr;
  }
};

export const cloneTransactionEntry = (entry) => {
  if (!entry) return null;
  const {
    _id: _unusedId,
    id: _unusedRawId,
    createdAt: _unusedCreated,
    updatedAt: _unusedUpdated,
    lastMutationId: _unusedMutId,
    lastMutationType: _unusedMutType,
    lastMutationAt: _unusedMutAt,
    lastEditedBy: _unusedEditedBy,
    ...rest
  } = entry;
  return {
    ...rest,
    date: new Date().toISOString(),
  };
};

File: native-android/app/src/main/java/com/paymatrix/app/data/LogTransactions.kt
package com.paymatrix.app.data

import com.paymatrix.app.domain.Money
import java.time.Instant
import java.util.Locale

data class LogCatalogItem(
    val id: String,
    val name: String,
    val kind: String,
    val archived: Boolean = false,
    val persisted: Boolean = false,
)

object LogDefaults {
    val accounts = listOf("Cash", "UPI", "Bank", "Credit Card", "Wallet").map {
        LogCatalogItem("account_${it.lowercase(Locale.ROOT).replace(" ", "_")}", it, "account")
    }
    val categories = listOf("Food", "Transport", "Shopping", "Entertainment", "Bills", "Health", "Education", "Travel", "Fuel", "Subscriptions", "Rent", "Groceries", "Other").map {
        LogCatalogItem("category_${it.lowercase(Locale.ROOT)}", it, "category")
    }
    val catalog = accounts + categories
}

data class LogTransactionDraft(
    val amount: String = "",
    val title: String = "",
    val category: String = "Other",
    val categoryId: String = "category_other",
    val accountId: String = "account_cash",
    val accountName: String = "Cash",
    val transactionType: String = "expense",
    val currency: String = "INR",
    val toAccountId: String = "",
    val toAccountName: String = "",
    val note: String = "",
    val place: String = "",
    val date: String = Instant.now().toString(),
    val friendId: String = "",
    val friendName: String = "",
    val sourceGroupId: String = "",
    val sourceGroupName: String = "",
    val sourceExpenseId: String = "",
) {
    fun fields(): Map<String, Any> {
        val paise = Money.toPaise(amount)
        require(paise in 1..100_000_000L) { "Enter an amount between ₹0.01 and ₹10,00,000." }
        require(currency == "INR") { "Only INR entry is supported at present." }
        require(transactionType in listOf("expense", "income", "transfer")) { "Choose a transaction type." }
        require(accountId.isNotBlank()) { "Choose an account." }
        require(transactionType == "transfer" || categoryId.isNotBlank()) { "Choose a category." }
        require(transactionType != "transfer" || (toAccountId.isNotBlank() && toAccountId != accountId)) { "Choose two different accounts." }
        Instant.parse(date)
        val resolvedTitle = title.trim().ifBlank { if (transactionType == "transfer") "Account transfer" else category }
        require(resolvedTitle.length in 1..100 && note.length <= 500 && place.length <= 100 && friendName.length <= 100) { "Shorten the title, note or reference." }
        return mapOf(
            "title" to resolvedTitle, "amountPaise" to paise, "amount" to paise / 100.0,
            "currency" to currency, "transactionType" to transactionType,
            "accountId" to accountId, "accountName" to accountName,
            "categoryId" to categoryId, "category" to category,
            "toAccountId" to if (transactionType == "transfer") toAccountId else "",
            "toAccountName" to if (transactionType == "transfer") toAccountName else "",
            "note" to note.trim(), "place" to place.trim(), "date" to date,
            "friendId" to friendId, "friendName" to friendName.trim(),
        )
    }

    companion object {
        fun from(entry: LogEntry, duplicate: Boolean = false) = LogTransactionDraft(
            amount = java.math.BigDecimal.valueOf(entry.amountPaise, 2).toPlainString(),
            title = entry.title, category = entry.category,
            categoryId = entry.categoryId.ifBlank { "category_${entry.category.lowercase(Locale.ROOT)}" },
            accountId = entry.accountId, accountName = entry.accountName,
            transactionType = entry.transactionType, currency = entry.currency,
            toAccountId = entry.toAccountId, toAccountName = entry.toAccountName,
            note = entry.note, place = entry.place,
            date = if (duplicate) Instant.now().toString() else runCatching { Instant.parse(entry.date).toString() }.getOrElse { java.time.LocalDate.parse(entry.date.take(10)).atStartOfDay(java.time.ZoneId.systemDefault()).toInstant().toString() },
            friendId = entry.friendId, friendName = entry.friendName,
            sourceGroupId = entry.sourceGroupId, sourceGroupName = entry.sourceGroupName, sourceExpenseId = entry.sourceExpenseId,
        )
    }
}
